### PR TITLE
Add GET /expenses with forward cursor pagination

### PR DESCRIPTION
**Pull request by mnovelodou**
## Summary

- Adds `GET /expenses` endpoint — the first read API for expenses
- Implements forward-only cursor pagination ordered by `expense_date DESC, expense_id DESC`
- `user_id` is a required query param (temporary until OAuth; will be inferred from token later)
- `start_date` / `end_date` are optional and independently defaulted; window capped at 3 calendar months
- Opaque base64 cursor encodes the last item's `(expense_date, expense_id)`; cursor date is validated to fall within the request window
- `limit` defaults to 20, max 100; rejects 0, negative, or above-cap values with 400
- Updated `idx_expenses_user_date` to `(created_by, expense_date DESC, expense_id DESC)` to cover the ORDER BY tie-breaker

## What changed

| Layer | Change |
|---|---|
| `schema.sql` | Index updated with `expense_id DESC` trailing column |
| `CursorPageResponse<T>` | New DTO: `content`, `nextCursor`, `pageSize` (no total count) |
| `ExpenseCursor` | Encode/decode utility for opaque base64 cursor tokens |
| `InvalidCursorException` | New exception → HTTP 400 via `GlobalExceptionHandler` |
| `ExpenseRepository` | `findByUserCursor()` — first page and cursor-page queries |
| `ExpenseService` | `listByUser()` — date defaults, validations, cursor handling |
| `ExpenseController` | `GET /expenses` handler with all query params |

## Date window defaulting

| Parameters | Resolved window |
|---|---|
| both absent | first–last day of previous calendar month |
| only `start_date` | `start_date` to `start_date + 1 month` |
| only `end_date` | `end_date - 1 month` to `end_date` |
| both present | used as-is; rejected if > 3 calendar months or end < start |

## Test plan

- [x] `ExpenseCursorTest` — encode/decode round-trip, URL-safe output, malformed input
- [x] `ExpenseServiceListByUserTest` — all date defaults, all validation branches, nextCursor logic
- [x] `ExpenseRepositoryTest` — first page, cursor page, empty result
- [x] `ExpenseControllerIT` — happy path, pagination flow, descending order, all 400 error cases

## Reviewer notes

- `user_id` is intentionally a query param for now; no auth layer exists yet. The service always applies it as `created_by` — no cross-user data exposure is possible.
- `nextCursor` is `null` (omitted in JSON) on the last page. If exactly `limit` items are returned there may be one extra empty page at the end — this is an acceptable trade-off vs. a `COUNT(*)` query.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/integrationTest/java/com/novelosoftware/expenses/controllers/ExpenseControllerIT.java [hunk 1]
```diff
@@ -22,6 +26,145 @@ void setUp() throws Exception {
...
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.content").isArray())
+            .andExpect(jsonPath("$.content.length()").value(2))
+            .andExpect(jsonPath("$.pageSize").value(20))
+            .andExpect(jsonPath("$.nextCursor").doesNotExist());
+    }
+
+    @Test
+    void list_defaultsToLastMonth_whenNoDatesProvided() throws Exception {
+        mockMvc.perform(get("/expenses")
```
Comment (mnovelodou) [inline src/integrationTest/java/com/novelosoftware/expenses/controllers/ExpenseControllerIT.java]: Can you check it actually returns the first page

File: src/main/java/com/novelosoftware/expenses/services/ExpenseService.java [hunk 2]
```diff
@@ -70,6 +77,85 @@ public UpdateExpenseResponse update(Long id, UpdateExpenseRequest request) {
...
+        if (userId == null || userId.isBlank()) {
+            throw createValidationException("user_id is required");
+        }
+
+        // --- Resolve date window ---
+        LocalDate resolvedStart = startDate;
+        LocalDate resolvedEnd = endDate;
+
+        if (resolvedStart == null && resolvedEnd == null) {
+            YearMonth lastMonth = YearMonth.now().minusMonths(1);
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/services/ExpenseService.java]: Lets initialize resolvedEnd with YearMonth.now() and then calculate resolvedStart with resolvedEnd.minusMonth(1)

File: src/test/java/com/novelosoftware/expenses/repositories/ExpenseRepositoryTest.java [hunk 3]
```diff
@@ -209,6 +209,50 @@ void get_notFound() {
         assertTrue(repo.get(id).isEmpty());
     }
 
+    // -------------------------------------------------------------------------
+    // findByUserCursor
+    // -------------------------------------------------------------------------
+
+    @Test
+    void findByUserCursor_firstPage_queriesWithoutCursorPredicate() {
+        when(jdbc.query(anyString(), any(RowMapper.class),
```
Comment (mnovelodou) [inline src/test/java/com/novelosoftware/expenses/repositories/ExpenseRepositoryTest.java]: Can you validate the SQL String used instead of using "anyString"

File: src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import java.time.LocalDate;
+import java.time.YearMonth;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class ExpenseServiceListByUserTest {
```
Comment (mnovelodou) [inline src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java]: Use existing test class for ExpenseService (ExpenseServiceTest)

File: src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    @InjectMocks
+    ExpenseService service;
+
+    // -------------------------------------------------------------------------
+    // Date defaults
+    // -------------------------------------------------------------------------
+
+    @Test
+    void listByUser_bothDatesAbsent_defaultsToLastCalendarMonth() {
```
Comment (mnovelodou) [inline src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java]: This test can be flakky, lets add a "Clock" dependency so we can mock now.

File: src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    void listByUser_limitZero_throws400() {
+        assertThrows(ExpenseValidationException.class,
+            () -> service.listByUser(USER, null, null, 0, null));
+    }
+
+    @Test
+    void listByUser_limitNegative_throws400() {
+        assertThrows(ExpenseValidationException.class,
+            () -> service.listByUser(USER, null, null, -5, null));
+    }
```
Comment (mnovelodou) [inline src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java]: These tests can be simplified with DataProviders.

File: src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        CursorPageResponse<Expense> result = service.listByUser(USER, null, null, null, null);
+
+        assertNotNull(result);
+        assertEquals(0, result.content().size());
+        assertNull(result.nextCursor());
+    }
+
+    @Test
+    void listByUser_onlyStartDateProvided_endDateDerivedAsStartPlusOneMonth() {
+        LocalDate start = LocalDate.of(2026, 4, 1);
```
Comment (mnovelodou) [inline src/test/java/com/novelosoftware/expenses/services/ExpenseServiceListByUserTest.java]: I was thinking that instead of testing all these combinations (defaulting start, end dates, cursor, etc) we should create a helper class that allow us to test this combinations separately as we will be using similar logic on other queries.

File: src/main/java/com/novelosoftware/expenses/controllers/ExpenseController.java [hunk 8]
```diff
@@ -1,36 +1,62 @@
...
+     * @param startDate optional start of date window; defaults to first day of last month
+     * @param endDate   optional end of date window; defaults to last day of last month
+     * @param limit     optional page size (1–100); defaults to 20
+     * @param cursor    optional opaque cursor from a previous response's {@code nextCursor}
+     * @return a page of expenses and an optional next-page cursor
+     */
+    @GetMapping
+    public CursorPageResponse<Expense> list(
+            @RequestParam(value = "user_id", required = false) String userId,
+            @RequestParam(value = "start_date", required = false)
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/controllers/ExpenseController.java:51]: Invalid `start_date`, `end_date`, or non-numeric `limit` values are converted by Spring before `listByUser` runs. Right now those conversion failures are not handled by `GlobalExceptionHandler`, so they fall through to the generic handler and return a 500 instead of the endpoint's expected 400. Please add coverage for malformed query params and handle `MethodArgumentTypeMismatchException` / conversion failures as `BAD_REQUEST`.

**Review by mnovelodou — COMMENTED** [review]
I left a few inline comments focused on behavior that can slip past the current tests: malformed query params, structurally incomplete cursors, and the index replacement path.

File: src/main/java/com/novelosoftware/expenses/util/ExpenseCursor.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+     * @return the decoded cursor position
+     * @throws com.novelosoftware.expenses.exceptions.ExpenseServiceExceptions.InvalidCursorException
+     *         if the token cannot be decoded or contains an invalid date
+     */
+    public static DecodedCursor decode(String token) {
+        try {
+            byte[] bytes = Base64.getUrlDecoder().decode(token);
+            Payload payload = MAPPER.readValue(bytes, Payload.class);
+            LocalDate date = LocalDate.parse(payload.d());
+            return new DecodedCursor(date, payload.id());
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/util/ExpenseCursor.java:60]: This accepts a token like base64(`{"d":"2026-05-01"}`): the date parses, `id` is null, and the repository query binds `expense_id < null`, returning an empty page instead of rejecting the malformed cursor. Please validate that `payload.id()` is present (and probably positive) before returning `DecodedCursor`, and add a test for a missing/null id.

File: src/main/resources/db/schema.sql [hunk 10]
```diff
@@ -26,7 +26,7 @@ CREATE TABLE IF NOT EXISTS expenses (
     FOREIGN KEY (account_id) REFERENCES accounts(account_id)
 );
 
-CREATE INDEX IF NOT EXISTS idx_expenses_user_date        ON expenses(created_by, expense_date);
+CREATE INDEX IF NOT EXISTS idx_expenses_user_date        ON expenses(created_by, expense_date DESC, expense_id DESC);
```
Comment (mnovelodou) [inline src/main/resources/db/schema.sql:29]: Because this keeps the existing index name with `IF NOT EXISTS`, databases that already created `idx_expenses_user_date` as `(created_by, expense_date)` will skip this statement and never get the new `expense_id DESC` tie-breaker. If this schema is applied incrementally anywhere, add an explicit drop/recreate or migration step so the index definition actually changes.

Comment (mnovelodou) [reply src/main/resources/db/schema.sql:29]: Resolving as Database still not exists in production, so it is safe to change.

File: src/main/java/com/novelosoftware/expenses/util/DateWindow.java [hunk 11]
```diff
@@ -0,0 +1,8 @@
+package com.novelosoftware.expenses.util;
+
+import java.time.LocalDate;
+
+/**
+ * A resolved, validated date window with an inclusive start and end date.
+ */
+public record DateWindow(LocalDate startDate, LocalDate endDate) {}
```
Comment (mnovelodou) [inline src/main/java/com/novelosoftware/expenses/util/DateWindow.java:8]: This record should be inside DateWindowResolver
